rocAL/source/readers/image/coco_file_source_reader.cpp

```diff
@@ -197,30 +228,40 @@ int COCOFileSourceReader::release() {
 
 void COCOFileSourceReader::shuffle_with_aspect_ratios() {
     // Calculate the mid element which divides the aspect ratios into two groups (<=1.0 and >1.0)
-    auto mid = std::upper_bound(_aspect_ratios.begin(), _aspect_ratios.end(), 1.0f) - _aspect_ratios.begin();
+    auto shard_start_idx = _shard_start_idx_vector[_shard_id];
+    auto mid = std::upper_bound(_aspect_ratios.begin() + shard_start_idx, _aspect_ratios.begin() + shard_start_idx + actual_shard_size_without_padding(), 1.0f) - (_aspect_ratios.begin() + shard_start_idx);
     // Shuffle within groups using the mid element as the limit - [start, mid) and [mid, last)
-    std::random_shuffle(_file_names.begin(), _file_names.begin() + mid);
-    std::random_shuffle(_file_names.begin() + mid, _file_names.end());
+    std::random_shuffle(_all_shard_file_names_padded.begin() + shard_start_idx, _all_shard_file_names_padded.begin() + shard_start_idx + mid);
+    std::random_shuffle(_all_shard_file_names_padded.begin() + shard_start_idx + mid, _all_shard_file_names_padded.begin() + shard_start_idx + actual_shard_size_without_padding());
     std::vector<std::string> shuffled_filenames;
-    int split_count = _file_names.size() / _batch_count;  // Number of batches for this shard
+    int split_count = (_all_shard_file_names_padded.size() /_shard_count) / _batch_size;  // Number of batches for this shard
     std::vector<int> indexes(split_count);
     std::iota(indexes.begin(), indexes.end(), 0);
     // Shuffle the index vector and use the index to fetch batch size elements for decoding
     std::random_shuffle(indexes.begin(), indexes.end());
     for (auto const idx : indexes)
-        shuffled_filenames.insert(shuffled_filenames.end(), _file_names.begin() + idx * _batch_count, _file_names.begin() + idx * _batch_count + _batch_count);
-    _file_names = shuffled_filenames;
+        shuffled_filenames.insert(shuffled_filenames.end(), _all_shard_file_names_padded.begin() + shard_start_idx + idx * _batch_size, _all_shard_file_names_padded.begin() + shard_start_idx + idx * _batch_size + _batch_size);
+    std::copy(_all_shard_file_names_padded.begin() + shard_start_idx, _all_shard_file_names_padded.begin() + shard_start_idx + actual_shard_size_without_padding(), std::back_inserter(shuffled_filenames));
 }
 
 void COCOFileSourceReader::reset() {
+    if (_stick_to_shard == false) // Pick elements from the next shard - hence increment shard_id
```

Comment:
Should we do this after shuffle or before that?
Please check file source reader and see which is the right way